lib/topology/platforms/shell.py

```diff
@@ -585,29 +594,54 @@ def connect(self, connection=None):
         )
 
         self._connections[connection] = spawn
+        index = None
 
         try:
+            # Setup shell before using it
+            self._pre_setup_shell(connection)
+
             # If connection is via user
             if self._user is not None:
-                spawn.expect(
-                    [self._user_match], timeout=self._timeout
+                index = spawn.expect(
+                    [self._user_match,
+                     self._initial_prompt,
+                     self._password_match],
+                    timeout=self._timeout
                 )
-                spawn.sendline(self._user)
+                if 0 == index:
+                    # Entering user name
+                    spawn.sendline(self._user)
+                elif 1 == index:
+                    # We already made authentication by passing user as
+                    # part of connect command: 'user@server'
+                    spawn.send('\n\r')
+                elif 2 == index:
+                    # User name was specified as part of part of connect
+                    # command 'user@server'. So, need to provide password
+                    spawn.sendline(self._password)
+                else:
+                    # We should not go here as timeout exception in this case
+                    # will be raised
+                    pass
 
             # If connection is via password
             if self._password is not None:
-                spawn.expect(
-                    [self._password_match], timeout=self._timeout
+                index = spawn.expect(
+                    [self._password_match, self._initial_prompt],
+                    timeout=self._timeout
                 )
-                spawn.sendline(self._password)
+                if 0 == index:
+                    spawn.sendline(self._password)
+                else:
+                    spawn.send('\n\r')
```

Comment:
This looks like you have a specific situation where a particular node that you are trying to connect to may be in an unknown state. Please handle specific situations in platform engine nodes, keep in mind that Topology is used for other nodes too that do not necessarily share the same requirements that you do.